### PR TITLE
Add proof of concept for T.Output()

### PR DESCRIPTION
**Pull request by suntala**
Addresses https://go.dev/issue/59928

### CODE REVIEW

File: src/testing/testing.go [hunk 1]
```diff
@@ -1034,6 +1034,52 @@ func (c *common) logDepth(s string, depth int) {
 	}
 }
 
+func (c *common) Output() io.Writer {
+	return OutputWriter{c}
+}
+
+type OutputWriter struct {
```
Comment (jba) [inline src/testing/testing.go]: unexport

File: src/testing/testing.go [hunk 2]
```diff
@@ -1034,6 +1034,52 @@ func (c *common) logDepth(s string, depth int) {
...
+	s.c.mu.Lock()
+	defer s.c.mu.Unlock()
+	if s.c.done {
+		// This test has already finished. Try and log this message
+		// with our parent. If we don't have a parent, panic.
+		for parent := s.c.parent; parent != nil; parent = parent.parent {
+			parent.mu.Lock()
+			defer parent.mu.Unlock()
+			if !parent.done {
+				parent.output = append(parent.output, parent.decorate(str, depth+1)...)
```
Comment (jba) [inline src/testing/testing.go]: This `Write` method is at a lower level than logDepth. It isn't responsible for adding line info or a final newline, so it shouldn't call `decorate`.  It does need to handle indentation, though. I'm not sure where that happens.

**Comment by jba** [comment]
I took another careful look at this. I do think you're on the right track: outputWriter.Write should buffer lines, direct the output to the right place (this test, parent test, etc.) and prepend four spaces. Here is what I think needs to be done:

- Move to gerrit so we can have a better review experience, others can comment, and your CL will be linked from the proposal issue. Gerrit also allows having multiple CLs stacked on top of one another, which I think will be valuable. See https://golang.org/doc/contribute.html.

- My gut tells me that you should have a function `outputWriter.writeLine([]byte)`. Ultimately all output is a line at a time, so that will be a useful building block. I think you might want it to rewrite the exising code (see below). If you want the arg to be `string` instead of `[]byte`, that's fine to start. I'm not sure if the arg should include the newline or not; you decide.

- Rewrite `logDepth` and `decorate` to use outputWriter. In fact, I think the first CL should do that rewrite and make sure all tests pass. It wouldn't even expose t.Output. It would just be an internal change. I think that would increase confidence that you're on the right track, and you could temporarily defer handling various edge cases.

- The second CL would export t.Output and add tests for things that can't happen with t.Log, which I think all have to do with partial lines. For example, what happens if you have an incomplete line and the test ends? You'll need to flush the buffer (adding a final newline).

**Comment by suntala** [comment]
@jba thank you so much for the guidance. Doing this in two CLs, with the first one being a rewrite of the log/logDepth is a good idea. Since your message, we did the refactor and got tests working. Next steps are going to be cleaning up the code and moving to gerrit. We're aiming to get them done in the next couple of days.

**Comment by jba** [comment]
Sounds good. I'll hold off reviewing until it's on gerrit.

**Comment by suntala** [comment]
https://github.com/golang/go/pull/73703 and https://github.com/golang/go/pull/71575 are merged.
